design-documents/redirect_domain_whitelist.md

```diff
@@ -0,0 +1,56 @@
+# Domain Whitelist for Frontend Redirects
+
+With the introduction of PWA as well as other server-client Magento solutions, the Magento host is acting as a headless server.
+This headless Magento server can serve multiple frontend applications, potentially on different domains.
+Currently there is no defined way that Magento knows which client domains it serves and are safe to redirect the user.
+This presents a problem for some PayPal payment solutions when trying to implement them through GraphQl.
+
+## Issue
+Several Paypal payment methods require some URL to redirect to once the payment has completed (e.g. [PayFlow](https://developer.paypal.com/docs/classic/payflow/integration-guide/#introducing-the-gateway-checkout-solutions)).
+Currently Magento uses the current host and some predefined controller that handles the redirect from PayPal.
+
+For the GraphQl implementation we want to support the scenario where a single "headless" Magento server will act as the API server for multiple frontend clients (possibly on different domains).
+In order to support this case, we need the ability for the frontend client to specify the URLs that PayPal will redirect the frontend to after the PayPal transaction.
+However, introducing the ability for the client to define the redirect URL opens a security concern.
+With an open redirect it could be possible for an attacker to manipulate this URL to send the user to a malicious website after payment.
+
+## Proposed Solution
+Currently, the Magento server has no way of knowing which domains all it's frontend clients are using.
```

Comment:
add the reason for why this is the solution: decoupled non hardcoded urls per store

Reply:
Can you clarify what you mean? I'm not really sure what to add.